### PR TITLE
Clean up coroutines implementation

### LINKED ISSUE #1584: Clean up coroutines implementation
The coroutines code needs some cleanup, pp constants like `HPX_HAVE_GENERIC_CONTEXT_COROUTINES` or `HPX_HAVE_COROUTINE_ARG_MAX` are not required and are currently just cluttering the code.

### PR DESCRIPTION
**Pull request by hkaiser**
This fixes #1584

### CODE REVIEW

File: hpx/util/coroutine/coroutine.hpp [hunk 1]
```diff
@@ -208,61 +205,14 @@ namespace hpx { namespace util { namespace coroutines
             std::move(target), id);
     }
 
-    //void reset()
-    //{
-    //    HPX_ASSERT(exited());
-    //    m_pimpl->reset();
-    //}
+    typedef typename arg_slot_traits::template at<0>::type arg0_type;
+    static const int arity = 1;
```
Comment (sithhell) [inline hpx/util/coroutine/coroutine.hpp]: Do we still need this?

File: hpx/util/coroutine/coroutine.hpp [hunk 2]
```diff
@@ -208,61 +205,14 @@ namespace hpx { namespace util { namespace coroutines
...
-    {
-      return call_impl_nothrow(arg_slot_type(
-        BOOST_PP_ENUM_PARAMS(HPX_HAVE_COROUTINE_ARG_MAX, arg)));
-    }
-
-#undef HPX_COROUTINE_PARAM_WITH_DEFAULT
-
-#else
+        typedef typename result_slot_traits::template at<0>::type arg0_type;
+        static const int arity = 1;
```
Comment (sithhell) [inline hpx/util/coroutine/coroutine.hpp]: same here.

File: hpx/util/coroutine/coroutine.hpp [hunk 3]
```diff
@@ -340,40 +288,8 @@ namespace hpx { namespace util { namespace coroutines
       return !empty() && !exited() && !waiting();
     }
 
-#if defined(HPX_HAVE_GENERIC_CONTEXT_COROUTINES)
```
Comment (sithhell) [inline hpx/util/coroutine/coroutine.hpp]: Does this mean we drop support for Boost.Context based coroutines?

Comment (hkaiser) [inline hpx/util/coroutine/coroutine.hpp]: We don't. As the ticket said (see #1584)

> the pp constant  HPX_HAVE_GENERIC_CONTEXT_COROUTINES  is currently used for two different things

This patch removes one of the use of `HPX_HAVE_GENERIC_CONTEXT_COROUTINES`

File: hpx/util/coroutine/coroutine.hpp (hunk 1 above)
Comment (hkaiser) [inline hpx/util/coroutine/coroutine.hpp]: It does not compile without those, if that's what you're asking.

File: hpx/util/coroutine/coroutine.hpp (hunk 2 above)
Comment (hkaiser) [inline hpx/util/coroutine/coroutine.hpp]: same here.

File: hpx/util/coroutine/coroutine.hpp (hunk 1 above)
Comment (sithhell) [inline hpx/util/coroutine/coroutine.hpp]: What I am trying to get at is all that's involved here still looks overly complicated and bloated for a single known signature: `R(A)`, if I am not mistaken, it just looks like there is still a lot of potential for more simplification (removing most of the traits etc.)

Comment (sithhell) [inline hpx/util/coroutine/coroutine.hpp]: But I guess this can still be cleaned up further some time in the future.

**Pull request merged**